// gitbook-plugin-craftml/index.js
var _ = require('lodash')
var path = require('path')

var tutorial = require('./tutorial')
module.exports = {
  website: {
    assets: "./book",
    js: [
      // "craftml.js"
    ],
    css: [
      "craftml.css",
      "https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.1.0/styles/default.min.css"
      // "https://cdnjs.cloudflare.com/ajax/libs/materialize/0.97.3/css/materialize.min.css"
      // "https://fonts.googleapis.com/icon?family=Material+Icons"
    ],
    html: {
      //"body:start": '<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.1.0/highlight.min.js"></script>',
    }
  },
  hooks:{
    "page:before": function(page) {
      if (page.content.match(/script\:tag/)){
        page.content = page.content.replace(/`script:tag`/g, "{{ 'script' | tag }}")
        return page
      }
      return page
    }
  },
  // Extend templating filters
  filters: {
    // // Author will be able to write "{{ 'test'|myFilter }}"
    t: function(s) {
        var p = path.dirname(this.ctx.file.path)
        var relative = path.relative(p, `language/transform/${s}.html`)
        return `<span class="doc-link transform"><a href="./${relative}" target="_doc"><code>${s}</code></a></span>`
    },
    l: function(s) {
        var p = path.dirname(this.ctx.file.path)
        var relative = path.relative(p, `language/layout/${s}.html`)
        return `<span class="doc-link layout"><a href="./${relative}" target="_doc"><code>${s}</code></a></span>`
    },
    attr: function(s) {
        var p = path.dirname(this.ctx.file.path)
        // console.log('s',s,'p',p)
        var relative = path.relative(p, `language/directive/${s}.html`)
        return `<span class="doc-link attr"><a href="./${relative}" target="_doc"><code>${s}</code></a></span>`
    },
    tag: tutorial.tag
  },
  blocks: {
    tutorial,
    craftml: require('./craftml')
  }
}
